src/frontend/xgboost_json.cc

```diff
@@ -23,16 +23,73 @@
 #include <iostream>
 #include <memory>
 #include <queue>
-#include <stack>
 #include <string>
 #include <utility>
 
 #include "xgboost/xgboost.h"
 
 namespace {
 
-template <typename StreamType>
-std::unique_ptr<treelite::Model> ParseStream(std::unique_ptr<StreamType> input_stream);
+template <typename StreamType, typename ErrorHandlerFunc>
+std::unique_ptr<treelite::Model> ParseStream(std::unique_ptr<StreamType> input_stream,
+                                             ErrorHandlerFunc error_handler);
+
+inline FILE* OpenFile(const char* filename) {
+#ifdef _WIN32
+  FILE* fp = std::fopen(filename, "rb");
+#else
+  FILE* fp = std::fopen(filename, "r");
+#endif
+  CHECK(fp) << "Failed to open file '" << filename << "': " << std::strerror(errno);
+  return fp;
+}
+
+class OwningFileReadStream {
```

Comment:
I'm not sure I'm fully understanding the point of wrapping `rapidjson::FileReadStream` in another layer of abstraction here. It seems like the only thing we get for the trouble is a logging call when the file is closed, and it requires those who know rapidjson to read the source code to understand what's going on with the wrapper. Is there something that I'm not seeing?

Reply:
The point of the wrapper is to auto-close the file pointer. For the error handling, we need to close the file and open it again.

Reply:
Actually I just found a way to remove the wrapper and still use error handling. See my latest commit.